cache/clock_cache.h

```diff
@@ -52,107 +54,173 @@ constexpr double kStrictLoadFactor = 0.7;  // See fast_lru_cache.h.
   };
-  uint8_t flags;
+
+  std::atomic<uint8_t> flags;
 
   enum ClockPriority : uint8_t {
-    NONE = (0 << kClockPriorityOffset),  // Not an element in the eyes of clock.
-    LOW = (1 << kClockPriorityOffset),   // Immediately evictable.
+    NONE = (0 << kClockPriorityOffset),
+    LOW = (1 << kClockPriorityOffset),
     MEDIUM = (2 << kClockPriorityOffset),
     HIGH = (3 << kClockPriorityOffset)
-    // Priority is NONE if and only if
+    // Priority LOW means immediately evictable, and HIGH means farthest from
+    // evictable. Priority is NONE if and only if:
     // (i) the handle is not an element, or
-    // (ii) the handle is an element but it is being referenced.
+    // (ii) the handle is an externally referenced element, or
+    // (iii) the handle is an element that was externally referenced but it's
+    //      not any more, and the clock_pointer_ has not swept through the
+    //      slot since the element stopped being referenced.
   };
 
-  // The number of elements that hash to this slot or a lower one,
-  // but wind up in a higher slot.
-  uint32_t displacements;
-
-  std::array<char, kCacheKeySize> key_data;
-
-  ClockHandle() {
-    value = nullptr;
-    deleter = nullptr;
-    hash = 0;
-    total_charge = 0;
-    refs = 0;
-    flags = 0;
-    SetIsVisible(false);
+  // The number of elements that hash to this slot or a lower one, but wind
+  // up in this slot or a higher one.
+  std::atomic<uint32_t> displacements;
+
+  // Synchronization rules:
+  // - Use a shared reference when we want the handle's identity
+  //    members (i.e., key_data, hash, value, IS_ELEMENT flag) to
+  //    remain untouched, but not modify them. The only updates
+  //    that a shared reference allows are:
+  //      * set CLOCK_PRIORITY to NONE;
+  //      * set the HAS_HIT bit.
+  //    Notice that these two types of updates are idempotent, so
+  //    they don't require synchronization across shared references.
+  // - Use an exclusive reference when we want identity members
+  //    to remain untouched, as well as modify any identity member
+  //    or flag.
+  // - displacements can be modified without holding a reference.
+  // - refs is only modified through appropiate functions to
+  //    take or release references.
+
+  ClockHandle()
+      : value(nullptr),
+        deleter(nullptr),
+        hash(0),
+        total_charge(0),
+        refs(0),
+        flags(0),
+        displacements(0) {
+    SetWillDelete(false);
     SetIsElement(false);
     SetClockPriority(ClockPriority::NONE);
     SetCachePriority(Cache::Priority::LOW);
-    displacements = 0;
     key_data.fill(0);
   }
 
-  Slice key() const { return Slice(key_data.data(), kCacheKeySize); }
-
-  // Increase the reference count by 1.
-  void Ref() { refs++; }
-
-  // Just reduce the reference count by 1. Return true if it was last reference.
-  bool Unref() {
-    assert(refs > 0);
-    refs--;
-    return refs == 0;
+  ClockHandle(const ClockHandle& other)
```

Comment:
where will the copy constructor and assignment be used?

Reply:
For instance, in ClockCacheShard::Release and ClockCacheShard::Erase we use the assignment operator, and in ClockCacheShard::EvictFromClock we use the copy ctor (when we push_back into deleted).